Collective lift and shift updates

jonrandahl opened the pull request:
This PR includes the following updates and resolves the linked issues:

- Refactors the elapsed time calculated for API requests to be resolved as microseconds rather than milliseconds. This is to improve the reporting of the elapsed time in the system tooling logs. [#104](https://github.com/epimorphics/hmlr-linked-data/issues/104)/[#111](https://github.com/epimorphics/hmlr-linked-data/issues/111)
- Minor text changes to the `Gemfile` to include instructions for running Epimorphics specific gems locally during the development of those gems.
- Updated the production `data_services_api` gem version to be at least the current version`~>1.3.2` (this is to cover out of sync release versions)
- Updated the production `lr_common_styles` gem version to be at least the current version `~>1.9.0` (this is to cover out of sync release versions)
- Refactored better guards in `entrypoint.sh` to ensure the required env vars are set accordingly or deployment will fail noisily.
- Refactored the error messages displayed to be semantically formatted as well as incorporated ALL lines for the returned message to be held in the variable instead of being possibly displayed outside of the intended context.
- Adjusted the processing of error status to use the application template; also includes adjustments to not double render the error response. [#103](https://github.com/epimorphics/hmlr-linked-data/issues/103)
- Added the updated configuration for the AWS credential improvements as per other readme's; included further info to run the application locally; resolved a markdown linting issue with using HTML in markdown.

jonrandahl committed:
build: updated version cadence 

Refactored the version cadence creation to include a SUFFIX value if provided; otherwise no SUFFIX is included in the version number.

jonrandahl committed:
styles: Updated rubocop settings

Expands Method block  length rule to allow for slightly longer method names

jonrandahl committed:
fix: Swap out Time.zone.now for Process.clock_gettime

Following community guidance for measuring elapsed time in rails the logging mechanism has been changed to use the CPU time instead of RealTime. This also has the added benefit to return the results in the requested `microseconds` unit without further mathematical manipulation.

jonrandahl committed:
build: Setting Epimorphics gems based on environment 

Refined text changes to the `Gemfile` to include instructions for running Epimorphics specific gems locally during the development of those gems. Updated the production environment gem versions to be at least the current version to cover out of sync release versions. Also includes slight reordering of specific groups of included gems for different environments.

jonrandahl committed:
docs: Updated CHANGELOG

jonrandahl committed:
feat: static asset cache headers

This sets the cache control headers for HMLR apps to be public and cacheable in the production environment. 
Price Paid Data uses a time limit of 5 minutes (300 seconds)

https://github.com/epimorphics/hmlr-linked-data/issues/114

jonrandahl committed:
build: No default values in production

In Production no default URL values are passed on the basis that missing configuration options represent a category of bug, and in that case the deployment should fail fast and noisily. This does include a default value of `true` for the `RAILS_SERVE_STATIC_FILES` nonetheless.

https://github.com/epimorphics/ijd-transfer-notes/blob/main/rails-app-pattern.md#environment-variables-and-run-time-configuration

jonrandahl committed:
styles: Better error page formatting

In order to improve the ux minor tweaks have been made to the layout and formatting of the returned error message and reference code presentation.

jonrandahl committed:
refactor: `RAILS_RELATIVE_URL_ROOT` renamed to `APPLICATION_PATH`

Following lengthy internal discussions the global env variable `RAILS_RELATIVE_URL_ROOT` has been renamed to `APPLICATION_PATH` to be more clear on it's use; needs to be monitored once deployed. Also includes better guards in `entrypoint.sh` to ensure the required env vars are set accordingly or deployment will fail noisily.

jonrandahl committed:
fix: better error control

Refactored the `StandardError` rescue block to use the provided status from the current exception outside of `MalformedSearchError` and `ArgumentError` types and otherwise fall back to a generic 500 status;

jonrandahl committed:
fix: better error message display

Refactored the error messages displayed to be semantically formatted as well as incorporated ALL lines for the returned message to be held in the variable instead of being possibly displayed outside of the intended context.

jonrandahl committed:
docs: Updated CHANGELOG 

Also includes minor update to layout format for older entries

jonrandahl committed:
docs: Updated CHANGELOG

andrew-pickin-epi committed:
Docker build optimizations

andrew-pickin-epi committed:
Merge branch 'spike/collective-lift-and-shift-updates' of github.com:epimorphics/ppd-explorer into spike/collective-lift-and-shift-updates

jonrandahl committed:
fix: Resolved Missing Template for errors

Adjusted the processing of error status to use the application template; also includes adjustments to not double render the error response.

https://github.com/epimorphics/hmlr-linked-data/issues/103

jonrandahl committed:
Merge remote-tracking branch 'origin/spike/collective-lift-and-shift-updates' into spike/collective-lift-and-shift-updates

jonrandahl committed:
docs: Updated CHANGELOG

jonrandahl committed:
ci: resolving gem locations

refactored the gem paths to use the remote Epimorphics repository locations, also includes the regenerated gemfile.lock reflecting as such.

jonrandahl committed:
docs: Updated README

Added the updated configuration for the AWS credential improvements as per other readme's; included further info to run the application locally; resolved a markdown linting issue with using HTML in markdown.

jonrandahl committed:
docs: Updated CHANGELOG

jonrandahl committed:
docs: Updated CHANGELOG

jonrandahl committed:
fix: resolve `RAILS_RELATIVE_URL_ROOT` env var

This resolves that Rails would want the `RAILS_RELATIVE_URL_ROOT` env var set, while using `APPLICATION_ROOT` is clearer in the `entrypoint.sh` code - this also updates the readme’s to reflect this change

jonrandahl committed:
docs: Updated CHANGELOG

Fixed typo for `APPLICATION_ROOT` env var

jonrandahl committed:
Merge branch 'dev-infrastructure' into spike/collective-lift-and-shift-updates

jonrandahl committed:
fix: resolve `RAILS_RELATIVE_URL_ROOT` env var

fixes variable for development environments

jonrandahl committed:
Merge remote-tracking branch 'origin/spike/collective-lift-and-shift-updates' into spike/collective-lift-and-shift-updates

jonrandahl committed:
docs: Updated README

bogdanadrianmarc reviewed (approved):
Looks good 👍

der commented on Dockerfile:
```diff
@@ -19,22 +19,23 @@ RUN apk add --update build-base
-RUN ./bin/bundle config set --local without 'development test' && \
-    ./bin/bundle install && \
-    mkdir log
+COPY .bundle/config /root/.bundle/config
```
What's the reason for dropping the previous `set --local without 'development test'`?  Just going by the name, with no knowledge of bundle config, sounds like you would want those omitted.

jonrandahl replied:
Thanks @der I'm actually investigating that at the moment as I noted massive differences with the dockerfiles in all the LR Apps...

jonrandahl replied:
> passing the --without flag to `bundle-install` prevents Bundler from installing certain groups specified in the `Gemfile`. Bundler persists this value  in  app/.bundle/config  so that calls to Bundler.setup do not try to find gems from the  Gemfile that you didn't install.  Additionally, subsequent calls to bundle install remember this setting and skip those groups. [^1]

So in this case it prevents the development and test grouped gems from installing unnecessarily. 

[^1]: https://manpages.ubuntu.com/manpages/bionic/man1/bundle-config.1.html

der reviewed (approved):
Looks fine, just one comment in line.

jonrandahl committed:
build: cleanup of build files for LR apps

Implements unification approach to Make, Docker and other associated build files to ensure the image process completes as expected.

jonrandahl committed:
fix: removed unused rescues

as these are not used in the app they were actually causing more issue on production images and therefore were removed.

jonrandahl committed:
refactor: rearranging code blocks

now mirrors other app structures

jonrandahl committed:
refactor: mirrored config setting across all apps

`config.assets.precompile` and `config.assets.version` have moved to `config/initializers/assets.rb`

jonrandahl committed:
build: unified environment configurations across all apps

Updated the matching configuration settings on each environment configuration file to uniify the settings for all apps

andrew-pickin-epi commented on Makefile:
```diff
@@ -15,7 +16,7 @@ API_SERVICE_URL?= http://localhost:8080
 COMMIT=$(shell git rev-parse --short HEAD)
 VERSION?=$(shell /usr/bin/env ruby -e 'require "./app/lib/version" ; puts Version::VERSION')
-TAG?=$(shell printf '%s-%s-%08d' ${VERSION} ${COMMIT} ${GITHUB_RUN_NUMBER})
+TAG?=$(shell printf '%s_%s_%08d' ${VERSION} ${COMMIT} ${GITHUB_RUN_NUMBER})
```
Has it been confirmed that changing the Tag format doesn't break deployment?

andrew-pickin-epi commented on README.md:
```diff
@@ -178,6 +234,9 @@ of the application:
 | name                       | description                                                          | typical value              |
 | -------------------------- | -------------------------------------------------------------------- | -------------------------- |
-| `RAILS_RELATIVE_URL_ROOT`  | The path from the server root to the application                     | `/app/ppd`                 |
+| `APPLICATION_ROOT`         | The path from the server root to the application                     | `/app/ppd`                 |
```
NO. This shouldn't be an environment variable unless the image has no knowledge of the path from build time. I VERY much suspect this is not the case.

What does "Typical Value" mean?

andrew-pickin-epi commented on entrypoint.sh:
```diff
@@ -5,27 +5,35 @@ set -e
 fi
 
-export RAILS_RELATIVE_URL_ROOT=${RAILS_RELATIVE_URL_ROOT:-'/app/ppd'}
+export RAILS_RELATIVE_URL_ROOT=${APPLICATION_ROOT:-'/app/ppd'}
```
No. Don't use environment variables.

andrew-pickin-epi commented on entrypoint.sh:
```diff
@@ -5,27 +5,35 @@ set -e
 export SCRIPT_NAME=${RAILS_RELATIVE_URL_ROOT}
 
-echo "{'ts': '`date -u +%FT%T.%3NZ`', 'message': {'text: 'Starting Standard Reports UI. API_SERVICE_URL=${API_SERVICE_URL} RAILS_ENV=${RAILS_ENV} RAILS_RELATIVE_URL_ROOT=${RAILS_RELATIVE_URL_ROOT}', 'level': 'INFO'}}"
+echo "{\"ts\": $(date -u +%FT%T.%3NZ), \"level\": \"INFO\", \"message\": \"Starting ${PUBLIC_NAME} application with RAILS_ENV=\"${RAILS_ENV}\", \"RAILS_RELATIVE_URL_ROOT\"=\"${RAILS_RELATIVE_URL_ROOT}\", \"SCRIPT_NAME\"=\"${SCRIPT_NAME}\"}"
```
No. Don't use environment variables.

der commented on entrypoint.sh:
```diff
@@ -5,27 +5,35 @@ set -e
 
-export RAILS_RELATIVE_URL_ROOT=${RAILS_RELATIVE_URL_ROOT:-'/app/ppd'}
+export RAILS_RELATIVE_URL_ROOT=${APPLICATION_ROOT:-'/app/ppd'}
 export SCRIPT_NAME=${RAILS_RELATIVE_URL_ROOT}
```
Is this needed? I thought we only used `RAILS_RELATIVE_URL_ROOT`

der reviewed (approved):
Looks fine to me apart from not understanding the point of `SCRIPT_NAME`.

Don't understand @andrew-pickin-epi 's concern about use of environment variable to set `RAILS_RELATIVE_URL_ROOT` but once that's resolved to mutual satisfaction I don't think I need to re-review so approving now.

jonrandahl committed:
Merge remote-tracking branch 'origin/dev-infrastructure' into spike/collective-lift-and-shift-updates

jonrandahl committed:
refactor: error handling now uses templates

improvements for handling error pages. will need to be ported onto the other apps as well post lift & shift testing

jonrandahl committed:
build: improvements to environment configs

jonrandahl committed:
build: streamlined entrypoint script

as per other HMLR apps

jonrandahl committed:
build: gemfile and lock updates 

specific to epimorphics internal gems

jonrandahl committed:
build: additional environment assets compilation

andrew-pickin-epi commented on CHANGELOG.md:
```diff
@@ -1,7 +1,34 @@
+- (Jon) Refactored the version cadence creation to include a SUFFIX value if
+  provided; otherwise no SUFFIX is included in the version number.
+- (Jon) Adjusted the processing of error status to use the application template;
+  also includes adjustments to not double render the error response.
```
APPLICATION_ROOT has been dropped.

andrew-pickin-epi commented on README.md:
```diff
@@ -191,3 +250,15 @@ Prometheus into `~/apps`, a starting command line would be something like:
+
+| name                       | description                                                          | typical value              |
+| -------------------------- | -------------------------------------------------------------------- | -------------------------- |
+| `APPLICATION_ROOT`         | The path from the server root to the application                     | `/app/ppd`                 |
```
APPLICATION_ROOT has been dropped.

andrew-pickin-epi commented on Makefile:
```diff
@@ -17,7 +18,7 @@ API_SERVICE_URL?=http://data-api:8080
 COMMIT=$(shell git rev-parse --short HEAD)
 VERSION?=$(shell /usr/bin/env ruby -e 'require "./app/lib/version" ; puts Version::VERSION')
-TAG?=$(shell printf '%s-%s-%08d' ${VERSION} ${COMMIT} ${GITHUB_RUN_NUMBER})
+TAG?=$(shell printf '%s_%s_%08d' ${VERSION} ${COMMIT} ${GITHUB_RUN_NUMBER})
```
This still using the  wrong path for the bundle conf.
Broken merge?

andrew-pickin-epi reviewed (changes requested):
I suspect there a broken merge going on here. 
The `Makefile` is certainly in error.

jonrandahl committed:
build: Cleaned BUNDLE_CONFIG

Removed `${HOME}/` from path

jonrandahl committed:
docs: Updated CHANGELOG

jonrandahl committed:
docs: Rewrote README

Includes further clarification on running the app individually, or as part of the HMLR suite of apps; alongside using Docker containers and accessing the Data API locally.

jonrandahl committed:
revert: rolled back tag separator rubocop suggestion
